includes/class.llms.user.permissions.php

```diff
@@ -55,19 +56,34 @@ public function __construct() {
 	 */
 	public function editable_roles( $all_roles ) {
 
-		$user = wp_get_current_user();
+		if ( is_multisite() && is_super_admin() ) {
+			return $all_roles;
+		}
 
-		$lms_roles = self::get_editable_roles();
+		$user       = wp_get_current_user();
+		$user_roles = $user->roles;
 
-		foreach ( $lms_roles as $role => $allowed_roles ) {
+		if ( in_array( 'administrator', $user_roles, true ) ) {
+			return $all_roles;
+		}
 
-			if ( in_array( $role, $user->roles, true ) ) {
+		$editable_roles = self::get_editable_roles();
 
-				foreach ( $all_roles as $the_role => $caps ) {
-					if ( ! in_array( $the_role, $allowed_roles, true ) ) {
-						unset( $all_roles[ $the_role ] );
-					}
-				}
+		$roles = array();
+		foreach ( $user_roles as $user_role ) {
+			if ( isset( $editable_roles[ $user_role ] ) ) {
+				$roles = array_merge( $roles, $editable_roles[ $user_role ] );
+			}
+		}
+
+		$roles = array_unique( $roles );
+
+		$llms_roles = array_keys( LLMS_Roles::get_roles() );
+
+		foreach ( array_keys( $all_roles ) as $role ) {
+			// only filter out LLMS roles.
+			if ( ! in_array( $role, $roles, true ) && in_array( $role, $llms_roles, true ) ) {
```

Comment:
I don't understand the purpose of this condition.
Say I'm an user with a custom role who can edit roles, and this role is 'my_custom_role' ok?
For me the variable `$roles` at line 79 above will be an empty array.
So: `! in_array( $role, $roles, true )` will be always `true`, because `$roles` is empty, then `in_array( $role, $llms_roles, true )` will be true only for the LifterLMS defined roles (`LLMS_Roles::get_roles()`). 
So this means that you're preventing 'my_custom_role' from changing the role of a Subscriber to Student (which is an LLMS_Role), for example.
We just don't need this behavior, why we should? Is there anything I'm missing?
What's your purpose here?

Reply:
This is to handle the case where the user has multiple LLMS roles that are defined in `LLMS_User_Permissions::get_editable_roles()`. Previously, if the user has both `lms_manager` and `instructor`, even though the `lms_manager` role should allow the user to edit `instructor`, `instructors_assistant`, `lms_manager`, `student`, the user will only be allowed to edit `instructors_assistant`, because it would remove all roles that are not mapped to `instructor` in `LLMS_User_Permissions::get_editable_roles()`.

With this change it would instead remove all LLMS roles that are not mapped to any of the roles that the user has that are defined in `LLMS_User_Permissions::get_editable_roles()`. This means that any roles that need to be able to edit LLMS roles (except Super Admin and `administrator`) must be configured in `LLMS_User_Permissions::get_editable_roles()`, eg.
```php
		$roles = apply_filters(
			'llms_editable_roles',
			array(
				'lms_manager' => array( 'instructor', 'instructors_assistant', 'lms_manager', 'student' ),
				'instructor'  => array( 'instructors_assistant' ),
                                'my_custom_role' => array( 'student' ),
			)
		);
```

Reply:
Ah I see, so I was missing the "configuration" idea.
Mmm I think that would be fine for users with multiple roles, when at least one of them is a key of the `LLMS_User_Permissions::get_editable_roles()` array.
I still don't think that if I'm a user with only one plenipotentiary role, my_custom_role, LifterLMS should forbid me to promote a Subscriber to Student unless I use the filter above.

So I think that if
`array_intersect_key( array_flip( $user->roles ), LLMS_User_Permissions::get_editable_roles() )` is `empty` we should return `$all_roles` very early.

Reply:
>So I think that if
array_intersect_key( array_flip( $user->roles ), LLMS_User_Permissions::get_editable_roles() ) is empty we should return $all_roles very early

Wouldn't that allow for example `student`s to edit all LLMS roles?

Reply:
I think that the best thing is that you test the code so that you're not forced to trust my words.
`student`s cannot edit users at all so they are not impacted by this piece of code.

Also, that method, as you can see, doesn't add anything to `$all_roles`, at most it removes some role so...

Reply:
Ok, so we allow all unmapped roles to edit all LLMS roles, and we rely on the edit user check to make sure that only specific roles can edit users

Reply:
yes, that's how it is in the current stable, and as far as I know, there's no need to alter this behavior.

> we rely on the edit user check to make sure that users with specific capabilities can edit users

For users who cannot edit other users the filter hook `'editable_role'` will just not be fired so...

Reply:
I've made the changes accordingly and updated the tests